sensing/autoware_pointcloud_preprocessor/src/outlier_filter/polar_voxel_outlier_filter_node.cpp

```diff
@@ -0,0 +1,1087 @@
+  auto polar_opt = extract_polar_from_xyz(iter_x, iter_y, iter_z);
+  bool is_primary = is_point_primary(*iter_return_type);
+  bool passes_intensity = meets_intensity_threshold(*iter_intensity);
+
+  // Step 2: Advance iterators (always advance, regardless of success/failure)
+  advance_cartesian_iterators(iter_x, iter_y, iter_z);
+  advance_intensity_return_iterators(iter_intensity, iter_return_type);
+
+  // Step 3: Early return on invalid coordinates
+  if (!polar_opt.has_value()) {
+    return std::nullopt;
+  }
+
+  // Step 4: Create voxel index and determine point classification
+  PolarVoxelIndex voxel_idx = polar_to_polar_voxel(*polar_opt);
+
+  return PointVoxelInfo{voxel_idx, is_primary, passes_intensity};
+}
+
+template <typename Predicate>
+PolarVoxelOutlierFilterComponent::VoxelIndexSet
+PolarVoxelOutlierFilterComponent::determine_valid_voxels_generic(
+  const VoxelPointCountMap & voxel_counts, Predicate predicate) const
+{
+  VoxelIndexSet valid_voxels;
+  for (const auto & [voxel_idx, counts] : voxel_counts) {
+    if (predicate(counts)) {
+      valid_voxels.insert(voxel_idx);
+    }
+  }
+  return valid_voxels;
+}
+
+bool PolarVoxelOutlierFilterComponent::is_point_primary(uint8_t return_type) const
+{
+  if (!use_return_type_classification_) {
+    return true;  // Treat all as primary in simple mode
+  }
+  auto it = std::find(
+    primary_return_types_.begin(), primary_return_types_.end(), static_cast<int>(return_type));
+  return it != primary_return_types_.end();
+}
+
+bool PolarVoxelOutlierFilterComponent::meets_intensity_threshold(uint8_t intensity) const
+{
+  return intensity <= intensity_threshold_;
+}
+
+PolarVoxelOutlierFilterComponent::ValidPointsMask
+PolarVoxelOutlierFilterComponent::create_valid_points_mask(
+  const PointVoxelInfoVector & point_voxel_info, const VoxelIndexSet & valid_voxels) const
+{
+  ValidPointsMask valid_points_mask(point_voxel_info.size(), false);
+
+  for (size_t i = 0; i < point_voxel_info.size(); ++i) {
+    if (is_point_valid_for_mask(point_voxel_info[i], valid_voxels)) {
+      valid_points_mask[i] = true;
+    }
+  }
+
+  return valid_points_mask;
+}
+
+bool PolarVoxelOutlierFilterComponent::is_point_valid_for_mask(
+  const std::optional<PointVoxelInfo> & optional_info, const VoxelIndexSet & valid_voxels) const
+{
+  if (!has_voxel_info(optional_info)) {
+    return false;
+  }
+
+  const auto & info = *optional_info;
+
+  if (!is_voxel_in_valid_set(info.voxel_idx, valid_voxels)) {
+    return false;
+  }
+
+  return passes_secondary_return_filter(info.is_primary);
+}
+
+bool PolarVoxelOutlierFilterComponent::has_voxel_info(
+  const std::optional<PointVoxelInfo> & optional_info) const
+{
+  return optional_info.has_value();
+}
+
+bool PolarVoxelOutlierFilterComponent::is_voxel_in_valid_set(
+  const PolarVoxelIndex & voxel_idx, const VoxelIndexSet & valid_voxels) const
+{
+  return valid_voxels.count(voxel_idx) > 0;
+}
+
```

Comment:
Same here, replacing the call to this function with `valid_voxels.count(voxel_index) > 0` seems easier.